glean-core/src/upload/mod.rs

```diff
@@ -140,6 +145,99 @@ impl PingUploadManager {
             None => PingUploadTask::Done,
         }
     }
+
+    /// Processes the response from an attempt to upload a ping.
+    ///
+    /// Spawns a thread for any disk IO necessary.
+    ///
+    /// Based on the HTTP status of said response,
+    /// the possible outcomes are:
+    ///
+    /// * **200 - 299 Success**
+    ///   Any status on the 2XX range is considered a succesful upload,
+    ///   which means the corresponding ping file can be deleted.
+    ///   _Known 2XX status:_
+    ///   * 200 - OK. Request accepted into the pipeline.
+    ///
+    /// * **400 - 499 Unrecoverable error**
+    ///   Any status on the 4XX range means something our client did is not correct.
+    ///   It is unlikely that the client is going to recover from this by retrying,
+    ///   so in this case the corresponding ping file can also be deleted.
+    ///   _Known 4XX status:_
+    ///   * 404 - not found - POST/PUT to an unknown namespace
+    ///   * 405 - wrong request type (anything other than POST/PUT)
+    ///   * 411 - missing content-length header
+    ///   * 413 - request body too large Note that if we have badly-behaved clients that
+    ///           retry on 4XX, we should send back 202 on body/path too long).
+    ///   * 414 - request path too long (See above)
+    ///
+    /// * **Any other error**
+    ///   For any other error, a warning is logged and the ping is re-enqueued.
+    ///   _Known other errors:_
+    ///   * 500 - internal error
+    ///
+    /// # Arguments
+    ///
+    /// `uuid` - The UUID of the ping in question.
+    /// `status` - The HTTP status of the response.
+    ///
+    /// # Return value
+    ///
+    /// Returns the JoinHandle for the spawned thread for test syncronization purposes.
+    ///
+    /// # Panics
+    ///
+    /// Will panic if unable to spawn a new thread.
+    pub fn process_ping_upload_response(&self, uuid: &str, status: u16) -> thread::JoinHandle<()> {
+        let mut task = "";
+        match status {
+            200..=299 => {
+                log::info!("Ping {} successfully sent {}.", uuid, status);
+                task = "delete";
+            }
+            400..=499 => {
+                log::error!(
+                    "Server returned client error code {} while attempting to send ping {}.",
+                    status,
+                    uuid
+                );
+                task = "delete"
+            }
+            _ => {
+                log::error!(
+                    "Server returned response code {} while attempting to send ping {}.",
+                    status,
+                    uuid
+                );
+                task = "re-enqueue";
+            }
+        };
+
+        // Spawn a new thread to perform the task defined above.
+        let local_queue = self.queue.clone();
+        let local_manager = self.directory_manager.clone();
+        let uuid = uuid.to_owned();
+        thread::Builder::new()
```

Comment:
Hm, if we continue with threading on the Rust side we eventually should keep a thread pool around (creating short-lived threads is not free).

One question: do we _need_ the thread here for I/O? Shouldn't this be off-thread already on the platform side?

Reply:
Yes, it should be called off the main thread. That is mostly the reason I asked if I should spawn a thread for it in the first place. It is not immediatelly obvious to me I should, since it is already off the main thread on the platform side.

@Dexterp37 advised me to spawn a thread for disk IO whenever possible. Could ou maybe elaborate more on why, Alessio?

Reply:
@badboy @brizental is this the case on all platforms? I'm sure that's the case for Android, but I'm not sure that's the case for Python/iOS.

If that's a problem for this PR, I'm fine with taking this to a separate one and use a thread-pool instead (which is a good suggestion regardless!)

Reply:
The upload manager runs on the separate concurrent queue on iOS.
Python is largely unsolved right now (there's an open bug how we want to handle it).

I'd be ok to simplify it here to not use the thread, file a bug and we can investigate/discuss our stance on doing threading more wide-spread in Rust

Reply:
All right, sounds good to me :) Let's at the very least add a comment here, explaining why we're ok for this being on the caller thread, putting in words the assumption of not being called on the main thread.

Let's also mention the known state of the other platforms.

Reply:
Ok, I filed a bug (https://bugzilla.mozilla.org/show_bug.cgi?id=1615597) and will add comments :)